src/python/pants/init/specs_calculator.py

```diff
@@ -125,6 +149,35 @@ def create(
                 filesystem_specs.append(FilesystemLiteralSpec(file_name))
             else:
                 address_specs.append(SingleAddress(address.spec_path, address.target_name))
+
+
+        if exprs:
+            # TODO: this should only be necessary for the `changed-since`/etc queries! This can be done by
+            # returning a dummy ScmWrapper if no `changed-*` queries are used!
+            scm = scm("The --query option")
+            # TODO(#7346): deprecate --owner-of and --changed-* in favor of --query versions, allow
+            # pipelining of successive query expressions with the command-line target specs as the initial
+            # input!
+            if len(exprs) > 1:
+                raise ValueError(
+                    "Only one --query argument is currently supported! Received: {}.".format(exprs)
+                )
+
+            scm_wrapper = UncachedScmWrapper.create(scm)
+            (expr_addresses,) = session.product_request(
+                QueryAddresses, [Params(scm_wrapper, exprs[0])]
+            )
+            logger.debug("expr addresses: %s", expr_addresses)
+            dependencies = tuple(
+                SingleAddress(a.spec_path, a.target_name) for a in expr_addresses.addresses
+            )
+            return Specs(
+                address_specs=AddressSpecs(
+                    dependencies=dependencies, exclude_patterns=exclude_patterns, tags=tags
+                ),
+                filesystem_specs=FilesystemSpecs(filesystem_specs),
+            )
```

Comment:
You'd want to move this up above all this SCM stuff, and after `if not changed_options.provided:`.

This file was refactored recently, and now it follows the high level flow of:

```
if not changed_options.provided:
  # assume it's direct file/target args, and return those directly
# all the SCM logic. Note that there's no else clause because the above `if` early returned.
```

I'm suggesting putting `if exprs:` in between those two parts.

Reply:
I refactored this file in between your comments -- please let me know if you like the result.